Add JavaScript test environment

sh19910711 opened the pull request:
This pull request is to add a JavaScript test environment. The main changes are as follow:

* Use Mocha to run JavaScript tests
* ~~Use CoffeeScript as a DSL only for writing JavaScript tests~~
  - ~~So, add `coffee-rails` to test dependency~~
  - ~~Let me know what do you think of this :)~~
* Add a fake rack middleware class to compile `.erb` ~~and `.coffee` (ref: 474acce25)~~
* Add a few tests with BDD style
* Add a rake task to run tests for the templates
* Run tests for the templates on Travis CI

The tests can be run on a web browser, and I've made sure that it works on Chrome 43 and Firefox 38.

#### Usage Example

Type the following command:

```shell
$ bundle exec rake test:templates
```

Otherwise:

```shell
$ bundle exec rake test:templates:serve
-> Access to http://localhost:29292/html/spec_runner.html from web browsers
```

![Screenshot](https://lh3.googleusercontent.com/wnA8sntEpxQQCvNjSh19UMhUEVM3oXi8uQbgWMs7E7I=w503-h377-no)

#### Related things

* PR #137
* rails/turbolinks#472

sh19910711 commented:
Hello, could you review this pull request?

Thank you.

gsamokovarov commented:
If we decided to ditch the CoffeeScript for the console code, let's ditch it for the tests too.

sh19910711 commented:
@gsamokovarov  Sure, I decided to go with the JavaScript for the console code, but please wait. I didn't decide to ditch the CoffeeScript, and I want to use it if possible.

In the last pull request, I totally agree with the team's views. However, keeping them in my mind, I think it's worth trying to use the CoffeeScript in this case. 

But yeah, this is an open source project, so I would like to hear your guys thoughts again.

Thanks.

sh19910711 commented:
@gsamokovarov Genadi, I am sorry to overlook the sentence *"Let's not take the bet at all and put the time somewhere else."* in [your comment](https://github.com/rails/web-console/pull/137#issuecomment-108202950).

OK, I've changed the CoffeeScript parts to JavaScript, and made sure it works on Chrome 43 and Firefox 38.

Thank you.

gsamokovarov commented:
Thank you Hiroyuki. For the tests, its not that big of a deal to be in CoffeeScript, but the consistency with the main JavaScripts are worth it, IMO as we don't have to context switch when we write application code and tests.

I like CoffeeScript too, but in the end of the day, the tests being in JavaScript is not so much hassle. Thank you for the work so far! I'll review the code during the weekend.

gsamokovarov commented on test/templates/config.ru:
```diff
@@ -0,0 +1,63 @@
+require_relative "fake_middleware"
+
+$current_dir = File.dirname(__FILE__)
```
No need for this to be a global. You can avoid the scope gates and use it as a top-level local. Also, look up the [`Pathname`](http://ruby-doc.org/stdlib-2.2.2/libdoc/pathname/rdoc/Pathname.html) class in the standard lib.

sh19910711 replied:
> Also, look up the Pathname class in the standard lib.

OK :)

> And another question. Do we need it to point here, or the project root. E.g. current_dir vs current_project, both of them are used below.

Let's use the latter one. I think it could make sense to reduce `../` for the paths. The `project_root` seems better as naming?

gsamokovarov replied:
Yup. Go for it. :+1:

gsamokovarov commented on test/templates/config.ru:
```diff
@@ -0,0 +1,63 @@
+map "/node_modules" do
+  node_modules = File.expand_path($current_dir + "/../../node_modules")
+  unless Dir.exists?(node_modules)
+    raise "missing the node_modules directory"
```
You can call `Kernel#abort` to do (almost) the same thing. Also `exit 1` is never called as noone is handling this exception. Ruby exits with a non-zero status code on exceptions anyway :)

gsamokovarov commented on test/templates/fake_middleware.rb:
```diff
@@ -0,0 +1,42 @@
+    raise "view_path() is not implemented"
+  end
+
+  def render(template)
```
Do you need all of those methods to be public?

gsamokovarov commented on test/templates/fake_middleware.rb:
```diff
@@ -0,0 +1,42 @@
+  end
+
+  def view_path
+    raise "view_path() is not implemented"
```
There is a `NotImplemented` error for this case, but I would just use the duck type here (just skip it and let Ruby raise). We won't need a really stable interface and don't have that many clients for it so we can help confused developers. :-)

gsamokovarov commented on test/templates/spec/dom_helpers_spec.js:
```diff
@@ -0,0 +1,126 @@
+var assert = chai.assert;
+
+describe("DOM helpers", function() {
+
```
:scissors:

gsamokovarov commented on test/templates/spec/dom_helpers_spec.js:
```diff
@@ -0,0 +1,126 @@
+describe("DOM helpers", function() {
+
+  describe("hasClass()", function() {
+
```
:scissors:

gsamokovarov commented on test/templates/spec/dom_helpers_spec.js:
```diff
@@ -0,0 +1,126 @@
+  });
+
+  describe("addClass()", function() {
+
```
:scissors:

gsamokovarov commented on test/templates/spec/dom_helpers_spec.js:
```diff
@@ -0,0 +1,126 @@
+  });
+
+  describe("removeClass()", function() {
+
```
:scissors:

gsamokovarov commented on test/templates/spec/dom_helpers_spec.js:
```diff
@@ -0,0 +1,126 @@
+    });
+  });
+});
+
```
:scissors:

gsamokovarov commented on test/templates/spec/repl_console_spec.js:
```diff
@@ -0,0 +1,8 @@
+
+describe("REPLConsole", function() {
+
+  it("pending");
```
No need to commit pending tests, commit them once you have them. We can merge this PR even without them and do them in a separate one.

sh19910711 commented:
OK, pushed updated version :)

@gsamokovarov Hi, could you please review it again? Thanks.

gsamokovarov commented on test/templates/spec/dom_helpers_spec.js:
```diff
@@ -0,0 +1,121 @@
+        beforeEach(function() {
+          removeClass(this.div, "hello");
+        });
+        it("should not have hello"); // hasClass() doesn't work if className is empty;
```
Is this context needed at al?

sh19910711 replied:
I will cut this context :scissors:

gsamokovarov commented:
Can we also make a rake task to reduce the `npm install` and rack server starting boilerplate? Thinking about it, testing with phantom js, so we can run the tests on Travis CI can be useful.

sh19910711 commented:
Pushed updated commits again :)

> Thinking about it, testing with phantom js, so we can run the tests on Travis CI can be useful.

:+1: x 9

> Can we also make a rake task to reduce the npm install and rack server starting boilerplate?

I want to try that.

sh19910711 commented on test/templates/html/spec_runner.html.erb:
```diff
@@ -0,0 +1,33 @@
+    <script src="/templates/console.js"></script>
+
+    <!-- find and load test cases -->
+    <% Dir.glob("spec/**/*_spec.*") do |spec| %>
```
I need to fix this glob path.

sh19910711 commented:
> Can we also make a rake task to reduce the npm install and rack server starting boilerplate?

I just tried writing it. Could you check the below links?

* Diff: https://github.com/sh19910711/web-console/compare/e3a9fa8412796d2b96ba0d1bddd47596c19a7112...57485d7d88db7ea7f30e741bb5f3e58f4be3c2bf
* Build log (travis-ci): https://travis-ci.org/sh19910711/web-console/jobs/66231024#L152

Thanks.

sh19910711 commented:
Oh yeah, I've updated the branch. I hope that it's getting better than before.

@gsamokovarov Could you please review the latest changes? JFYI, I changed the rake tasks and helpers.

Thanks.

gsamokovarov commented on test/templates/fake_middleware.rb:
```diff
@@ -0,0 +1,50 @@
+
+    # helper for test runners
+    def glob(glob_path)
+      root_path = Pathname.new(Helper.test_root)
```
Right now this class knows about the Helper. Can we drop this dependency somehow?

gsamokovarov commented on test/templates/config.ru:
```diff
@@ -0,0 +1,70 @@
+
+module Helper
+  def self.test_root
+    File.dirname(__FILE__)
```
Can you use `Pathname` here as well? Maybe even just extract a constant instead of module with class methods.

gsamokovarov commented on test/templates/config.ru:
```diff
@@ -0,0 +1,70 @@
+
+# test runners
+map "/html" do
+  class TestRunners < FakeMiddleware
```
I've been thinking, is the inheritance issued here the best tool we have? I think we can avoid it passing `req_path_regex`, `headers`, `view_paths` as initializer arguments. Something in the lines of:

```ruby
map "/html" do
  run FakeMiddleware.new(
    req_path_regex: %r{^/html/(.*)},
    headers: {"Content-Type" => "text/html"},
    view_paths: PROJECT_ROOT.join("html")
  )
end
```

This will require a small bit of changes to the `FakeMiddleware` class. It applies for all the instances we use subclassing in this file.

sh19910711 replied:
> I've been thinking, is the inheritance issued here the best tool we have? I think we can avoid it passing req_path_regex, headers, view_paths as initializer arguments. Something in the lines of:

Yeah, that's right. How about this commit?

https://github.com/sh19910711/web-console/commit/492a4b6823617024089560fa15dc82ef83bf9666

gsamokovarov commented on test/templates/fake_middleware.rb:
```diff
@@ -0,0 +1,50 @@
+
+  private
+
+  # default header
```
Is this comment really useful?

gsamokovarov commented on test/templates/config.ru:
```diff
@@ -0,0 +1,70 @@
+$:.push File.dirname(__FILE__)
+require "fake_middleware"
```
I think you can even define the `FakeMiddleware` class here and save the `LOAD_PATH` juggling.

gsamokovarov commented on test/templates/html/spec_runner.html.erb:
```diff
@@ -0,0 +1,35 @@
+    <script src="/templates/console.js"></script>
+
+    <!-- find and load test cases -->
+    <% glob("spec/**/*_spec.js") do |spec| %>
```
If you had the `PROJECT_ROOT` constant being a path_name, you could have probably dropped the whole helper here and used it instead.

gsamokovarov commented on Rakefile:
```diff
@@ -16,6 +16,43 @@ Rake::TestTask.new(:test) do |t|
+  namespace :templates do
+    task :all => [:npm, :rackup, :mocha, :kill]
+
+    # config
```
:scissors:

_(Comments like this don't give much value. As a rule of thumb, I always like to make my comments full sentences and try to say why to code was there to begin with, or something tricky about it. If I can't do that, I don't leave the comment, as it just creates noise to the code.)_

gsamokovarov commented on Rakefile:
```diff
@@ -16,6 +16,43 @@ Rake::TestTask.new(:test) do |t|
+    task :all => [:npm, :rackup, :mocha, :kill]
+
+    # config
+    work_dir    = File.dirname(__FILE__) + "/test/templates"
```
`Pathname` all the things :)

gsamokovarov commented on test/templates/fake_middleware.rb:
```diff
@@ -0,0 +1,50 @@
+  private
+
+  # default header
+  def header
```
I'd even name this `headers`.

gsamokovarov commented on Rakefile:
```diff
@@ -16,6 +16,43 @@ Rake::TestTask.new(:test) do |t|
 end
 
+namespace :test do
+  require "tmpdir"
```
Can you move this requires to the top of the file?

sh19910711 commented:
OK, I have pushed a new version again.

@gsamokovarov Thank you for taking your time and pointing out. Could you check the changes again? Here is [a diff](https://github.com/rails/web-console/compare/1f92effc7a9359e92eafd9651710363a4e3ac1d5...ab974fdc3a31e329f251321f6b21548b0f6cc21f) from last time.

 * * * * *

By the way, I changed `.travis.yml` not to run tests for templates on JRuby in the commit 31ef4245, because the `rackup` command seems to cause an error on JRuby.

Build Log: https://travis-ci.org/rails/web-console/jobs/66558451#L199

gsamokovarov commented:
> By the way, I changed .travis.yml not to run tests for templates on JRuby in the commit 31ef424, because the rackup command seems to cause an error on JRuby.

This is fine, we can only run them under CRuby, since its the JavaScript we care being tested.

gsamokovarov commented on Rakefile:
```diff
@@ -16,6 +18,39 @@ Rake::TestTask.new(:test) do |t|
+  namespace :templates do
+    task :all => [:npm, :rackup, :mocha, :kill]
+
+    work_dir    = Pathname(__FILE__).dirname + "test/templates"
```
`Pathname(__FILE__).dirname.join("test", "templates")`

gsamokovarov commented on Rakefile:
```diff
@@ -16,6 +18,39 @@ Rake::TestTask.new(:test) do |t|
+    task :all => [:npm, :rackup, :mocha, :kill]
+
+    work_dir    = Pathname(__FILE__).dirname + "test/templates"
+    pid_file    = Pathname(Dir.tmpdir) + "web_console.#{SecureRandom.uuid}.pid"
```
`Pathname(Dir.tmpdir).join("web_console.#{SecureRandom.uuid}.pid")`

gsamokovarov commented on Rakefile:
```diff
@@ -16,6 +18,39 @@ Rake::TestTask.new(:test) do |t|
+
+    task :npm do
+      Dir.chdir work_dir
+      system "npm install --silent"
```
`Dir.chdir(work_dir) { system "npm install --silent" }` so you preserve the current CWD.

sh19910711 replied:
> so you preserve the current CWD. Same for the instances below.

Ah yeah, I've got it!

sh19910711 commented:
I have pushed a new version, and added a "test:templates:serve" rake task.

@gsamokovarov Hello, could you review the latest changes?

 * * * * *

As a result, we can run tests for templates by the following command:

    $ bundle exec rake test:templates

Otherwise:

    $ bundle exec rake test:templates:serve
    => Access to "http://localhost:29292/html/spec_runner.html" from web browsers

Thank you :)

gsamokovarov commented on test/templates/config.ru:
```diff
@@ -0,0 +1,90 @@
+  def self.relative_path(path)
+    TEST_ROOT.join path
+  end
+  def self.root_path(path)
```
You just use this once, you can drop the method and the `Helper` module as a whole.

sh19910711 replied:
Exactly, I have just changed it.

gsamokovarov commented on test/templates/config.ru:
```diff
@@ -0,0 +1,90 @@
+end
+
+module Helper
+  def self.relative_path(path)
```
You can just use `TEST_ROOT.join` instead of `Helper.relative_path`.

gsamokovarov commented:
Thanks Hiroyuki!

sh19910711 commented:
@gsamokovarov Yeah, thank you very much for merging and so many reviews :) I appreciate your assists indeed, since they are very helpful and useful to me.

In addition, `Pathname` :clap:

Thank you.

gsamokovarov commented:
Ref. #121
